#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [14]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [15]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("data/train.csv")


## Data exploration

In [16]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [17]:
train.head()

laptop_ID Company            Product   TypeName  Inches  \
0       1104    Acer     Aspire ES1-523   Notebook    15.6   
1        114    Dell             XPS 13  Ultrabook    13.3   
2        172    Acer    Aspire A517-51G   Notebook    17.3   
3        918      HP      Elitebook 820    Netbook    12.5   
4        447  Lenovo  IdeaPad 320-15AST   Notebook    17.3   

                   ScreenResolution                         Cpu  Ram  \
0                          1366x768   AMD A8-Series 7410 2.2GHz  4GB   
1  Quad HD+ / Touchscreen 3200x1800  Intel Core i7 7560U 2.4GHz  8GB   
2       IPS Panel Full HD 1920x1080  Intel Core i5 8250U 1.6GHz  8GB   
3                 Full HD 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   
4                          1600x900   AMD A6-Series 9220 2.5GHz  8GB   

      Memory                           Gpu       OpSys  Weight  Price_euros  
0  500GB HDD                 AMD Radeon R5  Windows 10   2.4kg        387.0  
1  256GB SSD  Intel Iris Plus Graphics 640  Windows 10  1.23kg       1379.0  
2  256GB SSD          Nvidia GeForce MX150  Windows 10     3kg        854.0  
3  512GB SSD         Intel HD Graphics 620  Windows 10  1.26kg       1483.0  
4    1TB HDD                 AMD Radeon R4  Windows 10   2.8kg        519.0

## Data processing

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [18]:
train = train[["laptop_ID", "Inches","Price_euros"]]

In [19]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 2)

In [10]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [24]:
X_train

laptop_ID  Inches
25          26    15.6
84         776    15.6
10        1263    15.6
342        182    15.6
890        321    13.3
..         ...     ...
106        394    15.6
270        441    15.6
860        955    17.3
435        210    15.6
102        673    15.6

[729 rows x 2 columns]

### 3. Asignar el modelo (vacío) a una variable


In [25]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [26]:
model.fit(X_train, y_train)

LinearRegression()

In [28]:
pred = model.predict(X_test)
pred

array([1120.43868276, 1069.79534654, 1070.20842432, 1160.62136315,
       1208.45577045, 1022.56227854, 1154.97198544, 1077.94506983,
       1074.00873993, 1084.55546588, 1135.07084845, 1128.48391161,
       1102.85078136, 1054.32277823, 1131.84884174, 1137.90208507,
       1085.50151446, 1055.44096704, 1172.81694567, 1116.91773368,
       1150.18028315, 1160.67245885, 1083.34429783, 1091.0275446 ,
       1124.35313378, 1154.75565863, 1101.65688611, 1126.25329158,
       1072.19119768, 1119.5614315 , 1060.10126111, 1030.7727385 ,
       1176.36941461, 1144.34609849, 1117.30850377, 1081.93983336,
       1168.60355229, 1108.67920842, 1107.55986809, 1173.31263901,
       1169.59493897, 1106.80711591, 1139.93595413, 1140.711014  ,
       1173.89094791, 1060.22115413, 1083.09645116, 1136.38348519,
       1091.60585349, 1098.54556025, 1103.14857221, 1071.7047165 ,
       1089.12738679, 1130.05360719, 1137.13623737, 1103.42909025,
       1121.40200414, 1081.19629335, 1066.41732088, 1186.72787

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [31]:
from sklearn.metrics import mean_squared_error
import numpy as np
print("RMSE :", np.sqrt(mean_squared_error(y_test, pred)))

RMSE : 705.2026860694585


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [35]:
X_pred = pd.read_csv("data/test.csv")
X_pred.head()

laptop_ID   Company          Product            TypeName  Inches  \
0        750   Toshiba  Tecra X40-D-10G            Notebook    14.0   
1        726  Mediacom    Smartbook 142            Notebook    14.0   
2        633  Mediacom    SmartBook 140            Notebook    14.0   
3        363      Dell    Inspiron 7577              Gaming    15.6   
4        319      Asus    VivoBook Flip  2 in 1 Convertible    11.6   

                            ScreenResolution  \
0  IPS Panel Full HD / Touchscreen 1920x1080   
1                IPS Panel Full HD 1920x1200   
2                          Full HD 1920x1080   
3                          Full HD 1920x1080   
4                       Touchscreen 1366x768   

                                    Cpu  Ram              Memory  \
0            Intel Core i5 7200U 2.5GHz  8GB           256GB SSD   
1              Intel Atom Z8350 1.92GHz  4GB            32GB HDD   
2           Intel Atom x5-Z8350 1.44GHz  2GB  32GB Flash Storage   
3           Intel Core i5 7300HQ 2.5GHz  8GB             1TB HDD   
4  Intel Celeron Dual Core N3350 1.1GHz  2GB  32GB Flash Storage   

                       Gpu       OpSys  Weight  
0    Intel HD Graphics 620  Windows 10  1.25kg  
1    Intel HD Graphics 400  Windows 10   1.4kg  
2        Intel HD Graphics  Windows 10   1.4kg  
3  Nvidia GeForce GTX 1050  Windows 10  2.65kg  
4    Intel HD Graphics 500  Windows 10   1.1kg

In [36]:
X_pred = X_pred[["laptop_ID", "Inches"]]

In [37]:
X_pred

laptop_ID  Inches
0          750    14.0
1          726    14.0
2          633    14.0
3          363    15.6
4          319    11.6
..         ...     ...
386        693    15.6
387       1132    12.5
388         49    15.6
389       1243    15.6
390        924    12.5

[391 rows x 2 columns]

**¿Por qué da error?** 

In [38]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1094.83707236, 1092.854299  , 1085.17105224, 1097.8846358 ,
       1006.70009166, 1069.05180653, 1115.10019144, 1124.89992282,
       1097.63678913, 1150.59336094, 1068.1430354 , 1051.21605845,
       1112.59020488, 1085.4923023 , 1072.60427546, 1130.68301179,
       1027.22026957, 1086.98859448, 1110.35958485, 1147.28873867,
       1123.9177483 , 1125.50975157, 1173.23002346, 1169.01663007,
       1138.94456745, 1093.01953012, 1176.86510795, 1128.56652717,
       1087.88815344, 1131.4265518 , 1074.42181771, 1110.27696929,
       1149.10628092, 1159.59845661, 1147.53658534, 1057.70540997,
       1129.69162511, 1073.2417407 , 1080.62719662, 1075.88083243,
       1170.36999883, 1069.71273098, 1044.15645868, 1017.96732722,
       1071.199811  , 1119.52991164, 1075.49581995, 1148.55949188,
       1191.27173466, 1209.28192601, 1158.27660771, 1091.52323794,
       1090.11877347, 1164.22492778, 1071.7781199 , 1141.42303415,
       1098.71079136, 1060.73066571, 1101.40903944, 1075.55037

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [20]:
X_pred = pd.read_csv("Dataset/test.csv", index_col=0)
X_pred.head()

Company              Product   TypeName  Inches  \
id                                                     
1184      HP  Pavilion 15-BC000nv   Notebook    15.6   
815   Lenovo       Thinkpad T470s  Ultrabook    14.0   
1244      HP        EliteBook 840   Notebook    14.0   
1121      HP          Spectre Pro   Notebook    13.3   
443   Lenovo    IdeaPad 320-15IAP   Notebook    15.6   

                 ScreenResolution                                   Cpu  Ram  \
id                                                                             
1184  IPS Panel Full HD 1920x1080           Intel Core i7 6700HQ 2.6GHz  6GB   
815             Full HD 1920x1080            Intel Core i7 7500U 2.7GHz  8GB   
1244            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  4GB   
1121            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  8GB   
443             Full HD 1920x1080  Intel Pentium Quad Core N4200 1.1GHz  4GB   

         Memory                      Gpu       OpSys  Weight  
id                                                            
1184    1TB HDD  Nvidia GeForce GTX 960M  Windows 10  2.18kg  
815   256GB SSD    Intel HD Graphics 620  Windows 10  1.32kg  
1244  256GB SSD    Intel HD Graphics 520   Windows 7  1.54kg  
1121  256GB SSD    Intel HD Graphics 520  Windows 10  1.16kg  
443   128GB SSD    Intel HD Graphics 505  Windows 10   2.2kg

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [39]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1094.83707236, 1092.854299  , 1085.17105224, 1097.8846358 ,
       1006.70009166, 1069.05180653, 1115.10019144, 1124.89992282,
       1097.63678913, 1150.59336094, 1068.1430354 , 1051.21605845,
       1112.59020488, 1085.4923023 , 1072.60427546, 1130.68301179,
       1027.22026957, 1086.98859448, 1110.35958485, 1147.28873867,
       1123.9177483 , 1125.50975157, 1173.23002346, 1169.01663007,
       1138.94456745, 1093.01953012, 1176.86510795, 1128.56652717,
       1087.88815344, 1131.4265518 , 1074.42181771, 1110.27696929,
       1149.10628092, 1159.59845661, 1147.53658534, 1057.70540997,
       1129.69162511, 1073.2417407 , 1080.62719662, 1075.88083243,
       1170.36999883, 1069.71273098, 1044.15645868, 1017.96732722,
       1071.199811  , 1119.52991164, 1075.49581995, 1148.55949188,
       1191.27173466, 1209.28192601, 1158.27660771, 1091.52323794,
       1090.11877347, 1164.22492778, 1071.7781199 , 1141.42303415,
       1098.71079136, 1060.73066571, 1101.40903944, 1075.55037

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [40]:
sample = pd.read_csv("data/sample_submission.csv")

In [41]:
sample.shape

(391, 2)

In [24]:
sample.head()

id  Price_euros
0  1184         2158
1   815          511
2  1244         5290
3  1121          622
4   443         1140

In [25]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [42]:
submission = pd.DataFrame({"laptop_id": sample["laptop_ID"], "Price_euros": predictions_submit})

In [44]:
submission.head()

laptop_id  Price_euros
0        750  1094.837072
1        726  1092.854299
2        633  1085.171052
3        363  1097.884636
4        319  1006.700092

In [28]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [53]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_id.all() == sample.FWS.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [54]:
chequeator(submission)

You're ready to submit!
